### ISSUE TITLE
Bose Hubbard model in grand canonical ensemble

### ISSUE THREAD

**Issue by zejian-li** [netket/netket#148]
I tried modifying `netket/Tutorials/BoseHubbard1d/bosehubbard1d.py` by specifying a non-zero chemical potential, removing the `n_bosons` option and changing the sampler to `nk.sampler.MetropolisLocal`. (Code attached below.) However the optimization process doesn't seem to run correctly. The plot of cost function against #iteration looks like:
![figure_1](https://user-images.githubusercontent.com/24532267/51268577-6eeca900-19fb-11e9-850d-0069b644b588.png)
Am I missing something here?

```
import netket as nk
from mpi4py import MPI

# 1D Periodic Lattice
g = nk.graph.Hypercube(length=5, n_dim=1, pbc=True)

# Boson Hilbert Space
hi = nk.hilbert.Boson(graph=g, n_max=4)

# Bose Hubbard Hamiltonian
ha = nk.operator.BoseHubbard(U=1.0, hilbert=hi,mu=1.0)

# Jastrow Machine with Symmetry
ma = nk.machine.RbmSpinSymm(alpha=4, hilbert=hi)
ma.init_random_parameters(seed=1234, sigma=0.01)

# Sampler
sa = nk.sampler.MetropolisLocal(machine=ma)

# Stochastic gradient descent optimization
op = nk.optimizer.AdaMax()

# Variational Monte Carlo
vmc = nk.variational.Vmc(
    hamiltonian=ha,
    sampler=sa,
    optimizer=op,
    n_samples=1000,
    diag_shift=5e-3,
    use_iterative=False,
    method='Sr')

vmc.run(output_prefix='test', n_iter=4000)
```
Thanks!

**Comment by gcarleo** [comment]
Most likely it's because `MetropolisLocal` is a poor sampling choice when you have a chemical potential. One would need a custom sampler that both creates particles at random and also makes hoppings (like `MetropolisHop`) . On a small system you can use `ExactSampler` to see if the issue is due to the sampler.

**Comment by lazerjet** [comment]
> Most likely it's because `MetropolisLocal` is a poor sampling choice when you have a chemical potential. One would need a custom sampler that both creates particles at random and also makes hoppings (like `MetropolisHop`) . On a small system you can use `ExactSampler` to see if the issue is due to the sampler.

I tried on a small system with 3 sites and `n_max=3` with `ExactSampler` but the issue persists
```
# Bose Hubbard model created 
# U= 1
# V= 0
# mu= 1
# Nmax= 3
# RBM Initizialized with nvisible = 3 and nhidden = 12
# Symmetries are being used : 17 parameters left, instead of 51
# Exact sampler is ready 
# Adamax optimizer initialized with these parameters :
# Alpha = 0.001
# Beta1 = 0.9
# Beta2 = 0.999
# Epscut = 1e-07
# Using the Stochastic reconfiguration method
# Using Cholesky decomposition
# Variational Monte Carlo running on 4 processes

```
![exact](https://user-images.githubusercontent.com/24532267/51297972-9bd1a800-1a5d-11e9-808b-8dfa3a1eaf5f.png)

**Comment by gcarleo** [comment]
well, then you have to work harder to understand the origin of the problem :) the optimization pattern is very weird. Is that a reason why you are using AdaMax+Sr and not Sg+Sr?

**Comment by lazerjet** [comment]
Because it's less sensitive to hyperparameters :) But in fact I tried Sr with different learning rates and tuned `n_samples` as well, the cost function still shoots up.

**Comment by gcarleo** [comment]
@lazerjet , did you try also with other machines, for example the non symmetric ones? With larger local hilbert space, like for bosons, I expect that other non-binary wave-functions should do better (even though I am pretty sure that the fluctuations you see are due to the grand-canonical sampling being rather hard). Of course the already working alternative is to work in the canonical ensemble...

**Comment by VolodyaCO** [comment]
> Most likely it's because `MetropolisLocal` is a poor sampling choice when you have a chemical potential. One would need a custom sampler that both creates particles at random and also makes hoppings (like `MetropolisHop`) . On a small system you can use `ExactSampler` to see if the issue is due to the sampler.

How does MetropolisHop work? I'm trying also to reproduce a bose-hubbard system but with a site-dependent potential, and using MetropoilsHamiltonian, and I am getting bad results. Shall I open a new Issue?

**Comment by gcarleo** [comment]
```MetropolisHop``` moves a particle from a random site i to a random site j , provided that the occupation number on site j is smaller than the maximum occupation number allowed. This preserves the total number of particles, so this really is a way to simulate the model in the canonical ensemble. There is no need to open a new issue here, unless there is something not working in the code. This is most likely related to the choice of wave function and/or the sampling. The example we have on the bose hubbard is in the canonical ensemble, that should be a good starting point for changes. Sampling in the grand canonical requires some more thinking in terms of what moves one should consider, I guess...

**Comment by gcarleo** [comment]
Hi @zejian-li , not sure if you are still looking into this, however it seems to me that the ```RbmMultiVal``` (a one-hot encoded Rbm for non spin hilbert spaces) essentially solves this problem. 

Here is a script that works

```python
import netket as nk
from mpi4py import MPI
from scipy.sparse.linalg import eigsh

# 1D Periodic Lattice
g = nk.graph.Hypercube(length=5, n_dim=1, pbc=True)

# Boson Hilbert Space
hi = nk.hilbert.Boson(graph=g, n_max=4)

# Bose Hubbard Hamiltonian
ha = nk.operator.BoseHubbard(U=1.0, hilbert=hi,mu=1.0)

# Use scipy sparse diagonalization
vals, vecs = eigsh(ha.to_sparse(), k=1, which="SA")
print("eigenvalues with scipy sparse:", vals)

# RbmMultiVal with Symmetry
ma = nk.machine.RbmMultiVal(alpha=1, hilbert=hi,symmetry=True)
ma.init_random_parameters(seed=1234, sigma=0.01)

# Exact Sampler for testing 
sa = nk.sampler.ExactSampler(machine=ma)

# Stochastic gradient descent optimization
op = nk.optimizer.Sgd(0.01)

# Variational Monte Carlo
vmc = nk.variational.Vmc(
    hamiltonian=ha,
    sampler=sa,
    optimizer=op,
    n_samples=1000,
    diag_shift=0.1,
    use_iterative=True,
    method='Sr')

vmc.run(output_prefix='test', n_iter=4000)
```
![image](https://user-images.githubusercontent.com/28149892/80022339-e6847000-84a9-11ea-9947-c27dcd548021.png)

Notice that this is with the latest ```RbmMultiVal``` with symmetries, as implemented in this PR #385, bust most likely it should work also with the old version without symmetries. 
To scale this up you'd need to change the ExactSampler with a CustomSampler performing both hopping and annihilation/creation moves, otherwise MetropolisLocal will perform very poorly, as we were already discussing here before. 

I hope this helps.

**Comment by gcarleo** [comment]
@VolodyaCO I saw your paper on arXiv today, that's interesting. 

I was a bit puzzled though by this statement "However, we find that the technique is challenged whenever the system transitions between excitation manifolds, as the ground state is not learned correctly at these boundaries" since I can't reconcile it with the experiments I did (see also the plot above, or similar results you can easily obtain around the critical point U~3 mu=0). What is the parameter region where you find issues?

**Comment by VolodyaCO** [comment]
@gcarleo Glad that you find it interesting. Indeed, the goal of the study is to focus on the convergence of the state and how it reflects on the calculation of several observables. In general, the energy converges well, as seen in fig 3 of the paper (of course, had we used more hidden neurons, this convergence would have been even better). I have some plots of the energy as a function of VMC iterations, and it looks similar to the one you show above (Actually, I have like ~8000 of those plots). The energy converges to the one found with Lanczos diagonalisation almost always with very good accuracy. There are some exceptions, but only because sometimes the energy goes up as VMC advances. This might be due to the use of MetropolisLocal sampler, which is not the best sampler. However, given that the Hamiltonian preserves the number of excitations, we decided not to use the Hamiltonian sampler (which is a better sampler), because it samples on a fixed excitation manifold. Nonetheless, despite the use of MetropolisLocal, energy convergence is good.

As it is seen in the paper figures, our parameter region is the first and part of the second Mott lobes.  Also, figs 5 and 8 show more crudely that there are relatively important differences between the exactly calculated observables and the RBM calculated observables **at the boundaries of the MI-SF phases, and within the plateaus of the SF phase**. We also show that if we use more hidden units, the convergence is improved on those boundaries. Still, unfortunately, another region starts to be badly learnt (bottom-right part of fig5c, although we also show that this is probably due to noisy probabilities, which we can clean later on by performing a custom Hamiltonian sampling).

Our motivation is to use netket not to show that it approximates known results with high fidelity. We actually want to use netket to study open quantum systems that are of interest in quantum optomechanics in scenarios where it is impossible for us to compare netket results with traditional approaches because those conventional approaches blow up the memory usage that we can afford (1/2 TB). For this reason, we wanted first to test the method in the Bose-Hubbard model, mainly to start grasping how many neurons we should use for our other objectives (which we understand is an open question in deep learning in general).

**Comment by VolodyaCO** [comment]
@gcarleo Also, I will be cleaning and commenting on the code to make it public so that our experiments are entirely reproducible. We haven't submitted the paper yet, so if you have any comments, I'd be very glad!

**Comment by gcarleo** [comment]
@VolodyaCO great, it's always good to have open-source code that can be used to reproduce papers... 

In any case, my main comment is that difficulties such as "the energy goes up as VMC advances" are something that in this case are most likely **entirely due** to the choice of the sampling, and are unrelated with the ansatz itself. We have been discussing this already but MetropolisLocal is really a very bad choice for the grand-canonical ensemble, and it can lead to inaccurate estimates of both the gradients and the energy, thus learning wrong states!

So my suggestion, if you have time, is to consider redoing some of these calculations in the canonical ensemble with the the Hamiltonian sampler or use a CustomSampler that supports creation and annihilation of particles + random hoppings, in the grand-canonical ensemble.

If had to bet, I would say that the difference between observables that you see now at the boundary of the phase transitions will likely disappear with the correct sampler!

**Comment by VolodyaCO** [comment]
@gcarleo do you know where I can find the CustomSampler documentation? I see that it is no longer in netket.org

I agree that this is what might be happening: Near the boundary, it is difficult to "know" whether the system is in the MI or SF phase. Moreover, since MetropolisLocal is the most random of netket's samplers, it may propose Fock states that are related to a ground state in the MI or SF phase indiscriminately (this is confirmed in the tomography section of the paper), making the learnt ground state "noisier."  Maybe, this "noise" can be larger because of MetropolisLocal and possibly could diminish when using another sampler. However, for parameters well into the Mott lobes, or in the SF plateaus, MetropolisLocal works just fine. Furthermore, your argument states that MetropolisLocal can lead to inaccurate estimates of the gradients and the energy (compared with other samplers). I agree with that. But a phenomenon that is intriguing for me (and only now do I think about it thoroughly), is that if you take a look at the absolute error between the exact energy and the RBM state energy, this absolute error is minimum at those boundaries (fig 3d, note that I refer to the boundaries as the boundaries for 5 sites). Looking at fig 3d closely, the error is actually larger far from the boundaries. Therefore, at the boundaries, the energy convergence is best. Thus, it cannot be said that wrong states are learnt as a consequence of learning wrong energies. In fact, given that energy is better learnt at the boundaries, I would say that it is not the case that MetropolisLocal leads to inaccurate estimates of gradients or energy, for this case study.

**Comment by VolodyaCO** [comment]
@gcarleo Those ops are not stochastic matrices. I will try to build a python kernel that:
1. Picks a site at random and creates/destroys/ a boson or doesn't.
2. Swaps two random sites.

I will do it during the week and will let you know the results. What this sampler would do is to make it more probable to propose states that are more easily acceptable. A brute-force idea would be to use MetropolisLocal with a much larger number of samples (or to relax the acceptance ratio by doing |machine(new_state)/machine_old_state|^x, where x < 2). This would be ultimately equivalent to using the sampling that you propose (although the sampler you propose doesn't change the excitation manifold).
